File: OpenJava/FastaSequence.java


/******************************************************************************/
/**
  This class provides an object model for a FASTA sequence.
  @author	    Darrell O. Ricke, Ph.D.  (mailto: d_ricke@yahoo.com)
  Copyright:	Copyright (c) 2000 Paragon Software
  License:	    GNU GPL license (http://www.gnu.org/licenses/gpl.html)  
  Contact:   	Paragon Software, 1314 Viking Blvd., Cedar, MN 55011
 
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 2 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
public class FastaSequence extends Object
{


/******************************************************************************/

/** The sequence name */
private   String  name = "";

/** The sequence description */
private   String  description = "";

/** The sequence */
private   String  sequence = "";	// sequence

/** The sequence type */
private   byte  sequence_type = 0;


/******************************************************************************/
/** This constructor creates and initializes a new FastaSequence object. */
public FastaSequence ()
{
  initialize ();
}  // constructor FastaSequence


/******************************************************************************/
/** This method initializes the class variables. */
public void initialize ()
{
  name = "";
  description = "";
  sequence = "";
  sequence_type = 0;
}  // method initialize 


/******************************************************************************/
/** 
  This method returns the FASTA sequence description.

  @return String of current sequence description.
*/
public String getDescription ()
{
  return description;
}  // method getDescription


/******************************************************************************/
/**
  This method returns the current sequence length.

  @return int of current FASTA sequence length.
*/
public int getLength ()
{
  return sequence.length ();
}  // method getLength


/******************************************************************************/
/**
  This method returns the current FASTA sequence name.

  @return String of current FASTA sequence name.
*/
public String getName ()
{
  return name;
}  // method getName


/******************************************************************************/
/**
  This method returns the current FASTA sequence.

  @return String of current sequence.
*/
public String getSequence ()
{
  return sequence;
}  // method getSequence


/******************************************************************************/
/**
  This method returns the current sequence type.

  @return byte value indicating current sequence type.
*/
public byte getSequenceType ()
{
  return sequence_type;
}  // method getSequenceType


/******************************************************************************/
/** This method sets the current sequence name. */
public void setName ( String value )
{
  name = value;
}  // method setName


/******************************************************************************/
/** This method sets the current sequence description. */
public void setDescription ( String value )
{
  description = value;
}  // method setDescription


/******************************************************************************/
/** This method sets the current sequence. */
public void setSequence ( String value )
{
  sequence = value;
}  // method setSequence


/******************************************************************************/
/** This method sets the current sequence type. */
public void setSequenceType ( byte value )
{
  sequence_type = value;
}  // method setSequenceType


/******************************************************************************/
/* This method parses the a FASTA sequence header line. */
public void parseHeader ( String line )
{
  // Check for no header line.
  if ( line.length () <= 0 )
  {
    System.out.println ( "FastaSequence.parseHeader: *Warning* No header line." );
    return;
  }  // if

  // Check for invalid header line.
  if ( line.charAt ( 0 ) != '>' )
  {
    System.out.println ( "FastaSequence.parseHeader: *Warning* Invalid header line '" + line + "'" );
    return;
  }  // if

  int index1 = line.indexOf ( ' ' );
  int index2 = line.indexOf ( '\t' );
  if ( ( index2 > 0 ) && ( index2 < index1 ) )  index1 = index2;
  if ( ( index1 < 0 ) && ( index2 > 0 ) )  index1 = index2;

  if ( index1 < 0 )
  {
    name = line.substring ( 1 ).trim ();
    return;
  }  // if
 
  name = line.substring ( 1, index1 );

  if ( index1 + 1 < line.length () ) 
    description = line.substring ( index1 + 1 ).trim ();
}  // method parseHeader


/******************************************************************************/
/**
  This method returns a the current FASTA sequence as a text string.

  @return  String of current FASTA sequence and header line.
*/
public String toString ()
{
  StringBuffer str = new StringBuffer ( 10000 );
  str.append ( ">" + name + " " + description + "\n" );

  for ( int index = 0; index < sequence.length (); index += 50 )
  {
    if ( index + 50 >= sequence.length () )
      str.append ( sequence.substring ( index ) + "\n" );
    else
      str.append ( sequence.substring ( index, index + 50 ) + "\n" );
  }  // for

  return str.toString ();
}  // method toString


/******************************************************************************/

}  // class FastaSequence

File: OpenCpp/FastaSequence.cpp

#include <iostream>
#include <string>

using namespace std;


/**
  * This class implements a model for a FASTA sequence and header line.
  *
  * Author:     Darrell O. Ricke, Ph.D.  (mailto: d_ricke@yahoo.com)
  * Copyright:  Copyright (c) 2000 Darrell O. Ricke, Ph.D., Paragon Software
  * License:    GNU GPL license  (http://www.gnu.org/licenses/gpl.html)
  * Contact:    Paragon Software, 1314 Viking Blvd., Cedar, MN 55011
  *
  * This program is free software: you can redistribute it and/or modify
  * it under the terms of the GNU General Public License as published by
  * the Free Software Foundation, either version 2 of the License, or
  * (at your option) any later version.
  *
  * This program is distributed in the hope that it will be useful,
  * but WITHOUT ANY WARRANTY; without even the implied warranty of
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details.
  *
  * You should have received a copy of the GNU General Public License
  * along with this program.  If not, see <http://www.gnu.org/licenses/>. 
  */       
class FastaSequence
{
public:
  // Constructors
  explicit FastaSequence() { initialize(); }

  // This method returns the description of the FASTA sequence.
  string getDescription() { return description; }

  // This method returns the length of the FASTA sequence.
  int getLength() { return sequence.length(); }

  // This method returns the name of the FASTA sequence.
  string getName() { return name; }
 
  // This method returns the current FASTA sequence. 
  string getSequence() { return sequence; }
 

  //***************************************************************************
  // This method initializes the class FastaSequence. 
  void initialize() 
  {
    description = ""; 
    name = "";
    sequence = "";
  }  // method initialize


  //***************************************************************************
  // Parse the FASTA header line with sequence name and description.
  void parseHeader ( string line )
  {
    // Assert: valid line
    if ( line.size () <= 0 )  return;

    // Assert: header line starts with '>' character.
    if ( line.at ( 0 ) != '>' )  return;

    // Split the name from the description at the first white space.
    string::size_type index = line.find_first_of( " \t" );
    if ( index != string::npos )
    {
      name = line.substr ( 1, index - 1 );
      description = line.substr ( index + 1 );
    }  // if
    else
    {
      name = line.substr ( 1 );
    }  // else
  }  // method parseHeader


  //***************************************************************************
  // This method sets the description for the FASTA sequence.
  void setDescription ( string value ) { description = value; }


  //***************************************************************************
  // This method sets the name for the FASTA sequence.
  void setName ( string value ) { name = value; }


  //***************************************************************************
  // This method sets the FASTA sequence.
  void setSequence ( string value ) { sequence = value; }


  //***************************************************************************
  // This method converts this object to a FASTA sequence with header string.
  string toString()
  {
    string str = ">" + name + " " + description + "\n";

    for ( int index = 0; index < sequence.size(); index += 50 )
    {
      if ( index + 50 > sequence.size() )
	str += sequence.substr ( index ) + "\n";
      else
	str += sequence.substr ( index, 50 ) + "\n";
    }  // for

    return str;
  }  // method toString


//*****************************************************************************
protected:
  string name;		// FASTA sequence name

  string description;	// FASTA sequence description

  string sequence;	// FASTA sequence

};  // class FastaSequence


// Testing module.

//*****************************************************************************
// This main method tests the FastaSequence object.
/*
int main()
{
  FastaSequence fasta_seq;

  string header_line = ">Seq_Name This is the description.";
  fasta_seq.parseHeader ( header_line );

  cout << "Name: '" << fasta_seq.getName() << "'\n";
  cout << "Desc: '" << fasta_seq.getDescription() << "'\n";
};  // main
*/
